Upstream DynamoRIO i#7649 (commit 1be2930aa4, "AArch64 detach_state: Fix
sigreturn segfault"), adapted to release 11.3.0-1.

thread_set_self_mcontext() builds a signal frame on the stack and resumes
the thread with sigreturn. On AArch64 the frame's __reserved area was left
uninitialised and the SIMD context record list was only terminated on SVE
hardware. On non-SVE hardware (Apple Silicon) the kernel reads stack
garbage after the fpsimd record, rejects the frame and delivers SIGSEGV at
the resume pc. Seedbox hits this whenever a thread-synch-all (a JIT code
flush) resumes a thread that was waiting at a safe spot: the JVM crashes
in DynamoRIO's reset exit stub.
diff --git a/core/unix/include/sigcontext.h b/core/unix/include/sigcontext.h
--- a/core/unix/include/sigcontext.h
+++ b/core/unix/include/sigcontext.h
@@ -365,8 +365,8 @@
  * *_MAGIC constants, see below. The null terminator's magic number is 0.
  */
 struct _aarch64_ctx {
-    __u32 magic;
-    __u32 size;
+    __u32 magic; /* Magic number indicating the struct type. */
+    __u32 size;  /* Size of the context struct (including this header). */
 };
 
 #        define FPSIMD_MAGIC 0x46508001
diff --git a/core/unix/signal.c b/core/unix/signal.c
--- a/core/unix/signal.c
+++ b/core/unix/signal.c
@@ -3338,7 +3338,7 @@
 void
 thread_set_self_mcontext(priv_mcontext_t *mc, bool is_detach_external)
 {
-    kernel_ucontext_t ucxt;
+    kernel_ucontext_t ucxt = { 0 };
     sig_full_cxt_t sc_full;
     sig_full_initialize(&sc_full, &ucxt);
 #if defined(LINUX) && defined(X86)
diff --git a/core/unix/signal_linux_aarch64.c b/core/unix/signal_linux_aarch64.c
--- a/core/unix/signal_linux_aarch64.c
+++ b/core/unix/signal_linux_aarch64.c
@@ -282,34 +282,114 @@
 {
 #ifdef DR_HOST_NOT_TARGET
     ASSERT_NOT_REACHED();
-#endif
+#else
+#    define GET_RECORD_AT_OFFSET(byte_offset) \
+        (void *)((byte *)sc_full->fp_simd_state + byte_offset)
+
     /* sig_full_initialize() will have set the fp_simd_state pointer in the
      * user level machine context's (uc_mcontext) to __reserved.
      */
-    struct fpsimd_context *fpc = (struct fpsimd_context *)sc_full->fp_simd_state;
-    if (fpc == NULL)
+    size_t offset = 0;
+    struct _aarch64_ctx *next_head = GET_RECORD_AT_OFFSET(offset);
+
+    /* Make sure sc_full->fp_simd_state is not NULL. */
+    if (next_head == NULL) {
         return;
-    fpc->head.magic = FPSIMD_MAGIC;
-    fpc->head.size = sizeof(struct fpsimd_context);
-    fpc->fpsr = mc->fpsr;
-    fpc->fpcr = mc->fpcr;
-    ASSERT(sizeof(fpc->vregs) == (sizeof(mc->simd->q) * MCXT_NUM_SIMD_SVE_SLOTS));
-    int i;
-    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
-        memcpy(&fpc->vregs[i], &mc->simd[i].u32[0], sizeof(fpc->vregs[i]));
     }
 
-#ifndef DR_HOST_NOT_TARGET
-    if (proc_has_feature(FEATURE_SVE)) {
-        struct _aarch64_ctx *esr = (void *)((byte *)fpc + sizeof(struct fpsimd_context));
-        esr->magic = ESR_MAGIC;
-        esr->size = sizeof(struct esr_context);
+    struct fpsimd_context *fpsimd = NULL;
+    struct sve_context *sve = NULL;
 
-        struct sve_context *sve = (void *)((byte *)esr + sizeof(struct esr_context));
+    const struct _aarch64_ctx null_ctx = { 0 };
+
+    /* Search for the context records we are interested in. */
+    while (next_head->magic != 0) {
+        switch (next_head->magic) {
+        case FPSIMD_MAGIC: fpsimd = (void *)next_head; break;
+        case SVE_MAGIC: sve = (void *)next_head; break;
+        default: break;
+        }
+        offset += next_head->size;
+        next_head = GET_RECORD_AT_OFFSET(offset);
+    }
+
+    struct _aarch64_ctx *terminator_record = next_head;
+    ASSERT(terminator_record->magic == 0);
+
+    if (sve != NULL) {
+        /* Check that we have a full SVE record with space for the register data.
+         * The kernel only includes space for register data if it considers the SVE state
+         * to be "live" for the current thread.
+         * See https://docs.kernel.org/arch/arm64/sve.html for more information.
+         */
+        ASSERT(sve->vl == proc_get_vector_length_bytes());
+        if (sve->head.size < SVE_SIG_CONTEXT_SIZE(sve_vecquad_from_veclen(sve->vl))) {
+            /* sigcontext has an sve_context record but it doesn't contain any register
+             * data. We need to remove it and add a new larger record with space for the
+             * register data.
+             */
+            size_t sve_record_size = sve->head.size;
+
+            /* Shift the following records back on top of the SVE record. */
+            byte *dst = (void *)sve;
+            byte *src = dst + sve_record_size; /* start of the next revord  */
+            byte *end_of_records = (byte *)terminator_record + sizeof(*terminator_record);
+            size_t length = (ptr_uint_t)end_of_records - (ptr_uint_t)src;
+            memmove(dst, src, length);
+
+            /* Set sve to NULL so the code below will add a new sve_context record. */
+            sve = NULL;
+
+            /* And move next_head backwards to the new end of the list. */
+            offset -= sve_record_size;
+            next_head = GET_RECORD_AT_OFFSET(offset);
+            terminator_record = next_head;
+            ASSERT(terminator_record->magic == 0);
+        }
+    }
+
+    if (fpsimd == NULL) {
+        fpsimd = (struct fpsimd_context *)terminator_record;
+        fpsimd->head.magic = FPSIMD_MAGIC;
+        fpsimd->head.size = sizeof(struct fpsimd_context);
+
+        offset += fpsimd->head.size;
+        terminator_record = GET_RECORD_AT_OFFSET(offset);
+        *terminator_record = null_ctx;
+    }
+
+    if (sve == NULL && proc_has_feature(FEATURE_SVE)) {
+        sve = (struct sve_context *)terminator_record;
+        /* Set other fields of sve_context to zero. New fields may be added and
+         * unexpected values in those fields may cause problems. This is a small
+         * struct so the compiler will allocate it to a virtual register and
+         * optimise this initialisation.
+         */
+        const struct sve_context sve_zero = { 0 };
+        *sve = sve_zero;
         sve->head.magic = SVE_MAGIC;
         sve->vl = proc_get_vector_length_bytes();
         const uint quads_per_vector = sve_vecquad_from_veclen(sve->vl);
         sve->head.size = ALIGN_FORWARD(SVE_SIG_CONTEXT_SIZE(quads_per_vector), 16);
+
+        offset += sve->head.size;
+        terminator_record = GET_RECORD_AT_OFFSET(offset);
+        *terminator_record = null_ctx;
+    }
+
+    fpsimd->fpsr = mc->fpsr;
+    fpsimd->fpcr = mc->fpcr;
+    ASSERT(sizeof(fpsimd->vregs) == (sizeof(mc->simd->q) * MCXT_NUM_SIMD_SVE_SLOTS));
+    int i;
+    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
+        memcpy(&fpsimd->vregs[i], &mc->simd[i].u32[0], sizeof(fpsimd->vregs[i]));
+    }
+
+    if (proc_has_feature(FEATURE_SVE)) {
+        ASSERT(sve->vl == proc_get_vector_length_bytes());
+        const uint quads_per_vector = sve_vecquad_from_veclen(sve->vl);
+        ASSERT(sve->head.size >= SVE_SIG_CONTEXT_SIZE(quads_per_vector));
+
         for (uint i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
             memcpy((byte *)sve + SVE_SIG_ZREG_OFFSET(quads_per_vector, i),
                    &mc->simd[i].u32, sve->vl);
@@ -319,13 +399,6 @@
                    &mc->svep[i].u16, sve->vl / 8);
         }
         memcpy((byte *)sve + SVE_SIG_FFR_OFFSET(quads_per_vector), &mc->ffr, sve->vl / 8);
-
-        size_t offset = (proc_get_vector_length_bytes() * MCXT_NUM_SIMD_SVE_SLOTS) +
-            ((proc_get_vector_length_bytes() / 8) * MCXT_NUM_SVEP_SLOTS) + 16;
-        struct _aarch64_ctx *null =
-            (void *)((byte *)sve + sizeof(struct sve_context) + offset);
-        null->magic = 0;
-        null->size = 0;
     }
 #endif
 }
